core/trino-main/src/main/java/io/trino/execution/executor/dedicated/ConcurrencyController.java

```diff
@@ -19,15 +19,12 @@ class ConcurrencyController
 {
```

Comment:
```
A dynamic per-task controller calculates the desired concurrency
  based on output buffer utilization. We call this "target concurrency".
  The number of leaf drivers is capped by the lowest of target concurrency
  and the maximum number of drivers per task (controlled by task.max-drivers-per-task)
```

How does that relate to:

```
Once the minimum is satisfied, additional leaf drivers can get scheduled
  up to a maximum global target concurrency (controlled by task.min-drivers)
  Splits get chosen from all available tasks in a round-robin manner.
```
According to third point it seems that the only cap is `task.max-drivers-per-task` which is unbounded currently.

Do intermediate drivers take execution slot from leaf drivers (they shouldn't)?

Reply:
Each task has a limit given by min(target concurrency, max-drivers-per-task). There's a global limit controlled by task.min-drivers (once each task has the guaranteed minimum scheduled)

Reply:
See https://github.com/trinodb/trino/blob/5cfd8b6dfe8b5d254cee18530f3a00252da689cf/core/trino-main/src/main/java/io/trino/execution/executor/dedicated/ThreadPerDriverTaskExecutor.java#L184

Reply:
> Each task has a limit given by min(target concurrency, max-drivers-per-task).

So in practice zero size splits could make `target concurrency` unbounded?

Reply:
The desired concurrency per task, yes, but there’s an effective global limit given by min-drivers

Reply:
Thx. makes sense

Reply:
> but there’s an effective global limit given by min-drivers

so the "min" value acts as a "max"?

Reply:
Only on a per-task basis. Beyond the guaranteed minimum per task, it caps the drivers for a task based on how much other tasks are using. But if there are enough tasks running, the total number of drivers might be larger that this value -- hence why I call it "target" instead of "min" or "max".